### ISSUE TITLE
Let's encrypt from RSA to ECDSA message in logs - certificate not generated

### ISSUE THREAD

**Issue by jmcollin78** [home-assistant/addons#3380]
### Describe the issue you are experiencing

While trying to renew the certificate, I've got this message in logs: 
```
[10:45:47] INFO: Selected DNS Provider: dns-ovh
[10:45:48] INFO: Use propagation seconds: 60
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Are you trying to change the key type of the certificate named home.clouderial.fr from RSA to ECDSA? Please provide both --cert-name and --key-type on the command line to confirm the change you are trying to make.
Ask for help or search for solutions at https://community.letsencrypt.org. See the logfile /var/log/letsencrypt/letsencrypt.log or re-run Certbot with -v for more details.
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped
```

EDIT: the new certificates are not generated. This will cause an major issue because the site will not be visible (in one week for my case).

### What type of installation are you running?

Home Assistant OS

### Which operating system are you running on?

Home Assistant Operating System

### Which add-on are you reporting an issue with?

Almond

### What is the version of the add-on?

5.0.9

### Steps to reproduce the issue

1. just start the addon to generate a new certificate
2. look at the journal logs

## System Information

version | core-2023.12.4
-- | --
installation_type | Home Assistant OS
dev | false
hassio | true
docker | true
user | root
virtualenv | false
python_version | 3.11.6
os_name | Linux
os_version | 6.1.58-haos-raspi
arch | aarch64
timezone | Europe/Paris
config_dir | /config

<details><summary>Home Assistant Community Store</summary>

GitHub API | ok
-- | --
GitHub Content | ok
GitHub Web | ok
GitHub API Calls Remaining | 5000
Installed Version | 1.33.0
Stage | running
Available Repositories | 1373
Downloaded Repositories | 43

</details>

<details><summary>Home Assistant Cloud</summary>

logged_in | false
-- | --
can_reach_cert_server | ok
can_reach_cloud_auth | ok
can_reach_cloud | ok

</details>

<details><summary>Home Assistant Supervisor</summary>

host_os | Home Assistant OS 11.2
-- | --
update_channel | stable
supervisor_version | supervisor-2023.12.0
agent_version | 1.6.0
docker_version | 24.0.7
disk_total | 457.7 GB
disk_used | 18.3 GB
healthy | true
supported | true
board | rpi4-64
supervisor_api | ok
version_api | ok
installed_addons | Home Assistant Google Drive Backup (0.112.1), Samba share (12.2.0), InfluxDB (4.8.0), Glances (0.20.0), Let's Encrypt (5.0.9), NGINX Home Assistant SSL proxy (3.6.0), SQLite Web (4.0.0), AppDaemon (0.16.0), Piper (1.4.0), Whisper (1.0.0), Mosquitto broker (6.4.0), Zigbee2MQTT (1.33.0-1), Studio Code Server (5.14.2)

</details>

<details><summary>Dashboards</summary>

dashboards | 7
-- | --
resources | 24
views | 43
mode | storage

</details>

<details><summary>Recorder</summary>

oldest_recorder_run | 27 décembre 2023 à 21:17
-- | --
current_recorder_run | 2 janvier 2024 à 10:08
estimated_db_size | 570.98 MiB
database_engine | sqlite
database_version | 3.41.2

</details>

<details><summary>Sonoff</summary>

version | 3.5.3 (a8c6d45)
-- | --
cloud_online | 9 / 9
local_online | 9 / 9
debug | failed to load: 

</details>

### Anything in the Supervisor logs that might be useful for us?

```txt
-
```

### Anything in the add-on logs that might be useful for us?

```txt
[10:45:47] INFO: Selected DNS Provider: dns-ovh
[10:45:48] INFO: Use propagation seconds: 60
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Are you trying to change the key type of the certificate named home.clouderial.fr from RSA to ECDSA? Please provide both --cert-name and --key-type on the command line to confirm the change you are trying to make.
Ask for help or search for solutions at https://community.letsencrypt.org. See the logfile /var/log/letsencrypt/letsencrypt.log or re-run Certbot with -v for more details.
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped
```

**Comment by jmcollin78** [comment]
Switching from ECDSA to RSA in the configuration page have resolve the issue.
Maybe one explication line in the README should be important for others.

**Comment by burntoc** [comment]
I had this issue as well.  Toggling to rsa fixed it.  On a whim, I decided to change it back to ecdsa and tried again and it seems that works now, too.

**Comment by supersebbo** [comment]
> I had this issue as well. Toggling to rsa fixed it. On a whim, I decided to change it back to ecdsa and tried again and it seems that works now, too.

Interesting, I wonder why that works.  At a guess the latest version of Lets Encrypt defaults to ECDSA but won't auto convert unless explicitly instructed with the right parameter.  Probably the new type switch setting in the add-on doesn't get applied until you set it in the settings screen.

**Comment by W90W2A** [comment]
It shouldn't work for a key renewal.  cerbot requires explicit user inputs to change key type, it's called out in the code and has been this way since 2.0 changed the default cert to P-256:

`This function ensures that the user will not implicitly migrate an existing key
    from one type to another in the situation where a certificate for that lineage
    already exist and they have not provided explicitly --key-type and --cert-name.`

I've tried to toggle between RSA and ECDSA in the add-on to see if I could replicate what @burntoc observed and the add-on will not issue an ECDSA based cert no matter how many times I toggle back and forth.  Seems explicit migration code will need to be added for key renewals (or folks wanting to switch will have to figure out some workaround for the initial migration)

**Comment by burntoc** [comment]
I thought it was odd it would work as well.  I can only go by what the logs indicate, which lacked any errors, and the ceritificate issuance notification.  I'd guess more than likely it is still an RSA key used instead the indicated selection of ECDSA.

**Comment by hgy59** [comment]
> I had this issue as well. Toggling to rsa fixed it. On a whim, I decided to change it back to ecdsa and tried again and it seems that works now, too.

Are you sure it works now? IMHO you have to wait until the certificate is renewed to ensure it works.

**Comment by jmcollin78** [comment]
> Are you sure it works now? IMHO you have to wait until the certificate is renewed to ensure it works.

I have generated a new certificate with success after setting the option to rsa.

**Comment by burntoc** [comment]
You can force an update, IIRC.  In any case, I changed my cert from a hostname cert to wildcard at the same time.  The certificate info is showing the wildcard cert now, and the public key info shows Elliptical Curve (which is ECDSA), 256 bit.

**Comment by hgy59** [comment]
> > Are you sure it works now? IMHO you have to wait until the certificate is renewed to ensure it works.
> 
> I have generated a new certificate with success after setting the option to rsa.

my comment was related to the _change back to ecdsa_ that _works now too_

**Comment by matthewpucc** [comment]
As others are indicating, switching ecdsa -> rsa -> ecdsa does not work for me for renewals.

**Comment by joachimcarrein** [comment]
EDCSA gives the same notice for me.
Changing to RSA renews perfectly.

Changing back to EDCSA gives the same notice again.

**Comment by lukasmalkmus** [comment]
Was caught of guard here, as well. That new option was introduced and defaulted to.

**Comment by CaptainMalu** [comment]
Have the same problem. I wonder why I got some days in a row a notification from my automation, that the certificate was renewed and I have to restart Homeassistent.
I restarted and there where no new certificate.
I looked 3 or 4 times into the log of the addon until I finally spotted the notice.
I saw that the option is set to ecdsa but I thought: sure, why wouldn't I want ecdsa.

Would be nice to get an actual error message when the certificate couldn't be renewed and don't change options in an existing installation when it will lead to such errors.

**Comment by WillCodeForCats** [comment]
> It seems to me that the plugin should set the default type to ECDSA for new installations only - if the existing base is all RSA, it should remain RSA for them, since doing the upgrade could be breaking.

I actually had this discussion in PR #3130 where I submitted a PR to make ECDSA an option without changing RSA as the default. I thought that it would be better to make it an option first than flip the default without warning.

**Comment by thargy** [comment]
> Was caught of guard here, as well. That new option was introduced and defaulted to.

Me too, this is a nasty breaking change for anyone who installed Let's Encrypt with the recommended approach of running on a schedule.  I had no idea until I got an email from the Let's Encrypt Expiry Bot, and I think there will be plenty of others who fall into the trap of not realising something is wrong until they can't access their HA instance remotely...

+1 to please default the settings to the current key type for existing installs.

Also, does anyone have a walk-through for changing from RSA to ECDSA for those who want to take that approach?

**Comment by WillCodeForCats** [comment]
I reworked my previous PR as new PR #3424 if anyone wants to look at or test my proposed solution to this issue.

**Comment by spali** [comment]
Had the same this morning... could find a way to set config to rsa on the shell, as I didn't had access to the UI. Had to disable ssl in configuration.yaml on the command line and restart core. Then change settings in the UI to rsa and restarted the addon.
Then changing config back to ssl and it worked.
Is there a path to switch the cert to ecdsa without hacking around?

**Comment by WillCodeForCats** [comment]
> Is there a path to switch the cert to ecdsa without hacking around?

It should be fixed in my PR.

**Comment by hgy59** [comment]
regarding the error message:
```
Are you trying to change the key type of the certificate named home.clouderial.fr from RSA to ECDSA? Please provide both --cert-name and --key-type on the command line to confirm the change you are trying to make.
```

The text `Please provide both --cert-name and --key-type on the command line to confirm the change you are trying to make.` indicates that `--key-type` parameter is not added with the default key type.

IMHO this would be fixed when no default keytype is used and `--key-type` is always added to the command line (to call certbot?).

I may be wrong and the `--key-type` argument is always applied.

There are some informations available in the discussion https://community.letsencrypt.org/t/existing-rsa-key-cert-renewed-as-ecdsa-on-certbot-upgrade/195317

especially the Posting [#5](https://community.letsencrypt.org/t/existing-rsa-key-cert-renewed-as-ecdsa-on-certbot-upgrade/195317/5)

**Comment by tyjtyj** [comment]
Hope this will be listed as breaking change. Caught me off guard and getting cert expiry alert leads me here. This is my wildcard cert thus changing with ECDSA might break other application that rely on this cert. 
```
# Changelog

## 5.0.9

- Add option to specify Private Key type
```

**Comment by Lectere** [comment]
This is a breaking change. This destroyed my home assistant setup. And it got destroyed by a update. This is severe. 

Does anyone have a thread on this on Home Assistant forums?, I could not find it...

Thanks OP for reporting this.

Been saying for years now that the guy's at Home Assistant should incorporate proper certificate management from the GUI of Home Assistant itself. You cannot expect people to setup reverse proxies and as it seems you cannot rely on third party tools like LetsEncrypt. Home Assistant is getting to important to be wracked by stupid issues like this...

**Comment by MZorzy** [comment]
but have not this issue with duckdns addon.

**Comment by lukasmalkmus** [comment]
> _[...]_ as it seems you cannot rely on third party tools like LetsEncrypt.

This has nothing to do with Let's Encrypt, the add-on just made a breaking change without a major version bump or any notification, whatsoever.

> You cannot expect people to setup reverse proxies _[...]_

There is the option to pay for a subscription if you don't want to deal with the downsides of hosting software yourself.

> This destroyed my home assistant setup. And it got destroyed by an update. This is severe.

Let's Encrypt sends out an email that certificates are about to expire. The Add-on won't rotate valid certificates out unless it receives valid new ones. So you definitely missed something.

Also there is the option to disable auto-updates for Add-ons.

This is OSS. There are things that shouldn't happen but sometimes they still do. This was a wakeup call. Hope next time more attention is paid when addressing this Add-on.

**Comment by WillCodeForCats** [comment]
These renewal issues and unexpected certificate key type changes should all be fixed in release 5.0.13.

The `keytype` setting introduced in 5.0.9 was removed. The default will be to renew existing certificates using the same key type as they were before, and create new certificates with the Certbot default (ECDSA). You can still set the key type instead of using auto detect with the new optional config value `key_type` to change between key types.

If there are still any ongoing problems with certificate renewal let me know and I take another look.

**Comment by spali** [comment]
Thanks for the work. 👍 
Worked like a charm for my case (had rsa):

1. updated addon to 5.0.13
2. Removed `keytype` (yaml)
3. Set new `key_type` to ECDSA
4. started addon (got new certificate with ECDSA)
5. removed explicit new `key_type` (yaml) to make it as new default
6. started addon (cert did not renew as expected)
7. restart homeassistant to pick up the new cert.

**Comment by WillCodeForCats** [comment]
When auto detect is working (no `key_type` setting) it will log:

`INFO: Detecting existing certificate type for X`
`INFO: Existing certificate using '(ecdsa|rsa)' key type.`

If `key_type` is set there won't be any extra logging.

**Comment by WillCodeForCats** [comment]
> but have not this issue with duckdns addon.

The duckdns addon is actually using [dehydrated](https://github.com/dehydrated-io/dehydrated), not Certbot like this one is using.

**Comment by agners** [comment]
@WillCodeForCats I've run the new add-on in my production instance and run into this error:

```
[09:19:26] INFO: Detecting existing certificate type for homeassistant.my.domain.tld
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Renewal configuration file /data/letsencrypt/renewal/homeassistant.my.domain.tld.conf produced an unexpected error: expected /data/letsencrypt/live/homeassistant.my.domain.tld/cert.pem to be a symlink. Skipping.
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Requesting a certificate for homeassistant.my.domain.tld and mqtt.my.domain.tld

Successfully received certificate.
Certificate is saved at: /data/letsencrypt/live/homeassistant.my.domain.tld-0003/fullchain.pem
Key is saved at:         /data/letsencrypt/live/homeassistant.my.domain.tld-0003/privkey.pem
This certificate expires on 2024-04-25.
```

The old certificate was an RSA, the new one is ECDSA. It seems that the `certbot certificates` command fails in my setup.

`ls -la /data/letsencrypt/live/homeassistant.my.domain.tld` shows that those are indeeed actual files instead of symlinks :man_shrugging: 

I have a very similar test setup, which I have not updated yet. I've run the same command there too, and it worked out just fine.

So it seems the live database somehow corrupted? :thinking: I don't remember messing with anything in there. I wonder if this is because it is a very old install which started off with a very early add-on version :man_shrugging:

**Comment by WillCodeForCats** [comment]
I assume Certbot has a filesystem layout it manages, but to be honest I don't anything about the inner workings of certbot. I moved away from it some years ago and replaced it with dehydrated because certbot was getting too top heavy.

To me that error looks like the older version of certbot didn't use symlinks to point to the current certificate, but now the newer version does that without backwards compatibility if it isn't.

I found references to run `certbot update_symlinks` but that's also a depreciated command and removed in 2.5.0.
